module/zfs/dmu_send.c

```diff
@@ -1537,7 +1541,8 @@ restore_object(struct restorearg *ra, objset_t *os, struct drr_object *drro)
 		}
 	}
 
-	tx = dmu_tx_create(os);
+	count = DN_BONUSLEN_TO_COUNT(drro->drr_bonuslen);
+	tx = dmu_tx_create_dn_count(os, count);
```

Comment:
Why does it matter if the preferred dnode size changes mid-way through the transaction?  I think the ZPL should decide what it's going to do, and then tell the DMU with dmu_tx_hold_* and dmu_object_alloc().